cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,321 @@
+  private final TopicIdMetaKey[] metricsTopicMetaKeys;
+  private final MessagingService messagingService;
+  private final DatumReader<MetricValues> recordReader;
+  private final Schema recordSchema;
+  private final MetricStore metricStore;
+  private final int fetcherLimit;
+
+  private long lastLoggedMillis;
+  private long recordsProcessed;
+
+  private Map<String, String> metricsContextMap;
+  private List<ProcessMetricsThread> processMetricsThreadList;
+
+  @Inject
+  public MessagingMetricsProcessorService(MetricDatasetFactory metricDatasetFactory,
+                                          @Named(Constants.Metrics.TOPIC_PREFIX) String topicPrefix,
+                                          @Assisted Set<Integer> topicNumbers,
+                                          MessagingService messagingService,
+                                          SchemaGenerator schemaGenerator,
+                                          DatumReaderFactory readerFactory,
+                                          MetricStore metricStore,
+                                          @Named(Constants.Metrics.MESSAGING_FETCHER_LIMIT)
+                                            int fetcherLimit) {
+    super(metricDatasetFactory);
+    this.metricsTopics = new TopicId[topicNumbers.size()];
+    this.metricsTopicMetaKeys = new TopicIdMetaKey[topicNumbers.size()];
+    int i = 0;
+    for (int topicNum : topicNumbers) {
+      this.metricsTopics[i] = NamespaceId.SYSTEM.topic(topicPrefix + topicNum);
+      this.metricsTopicMetaKeys[i] = new TopicIdMetaKey(this.metricsTopics[i]);
+      i++;
+    }
+    this.messagingService = messagingService;
+    try {
+      this.recordSchema = schemaGenerator.generate(MetricValues.class);
+      this.recordReader = readerFactory.create(TypeToken.of(MetricValues.class), recordSchema);
+    } catch (UnsupportedTypeException e) {
+      // This should never happen
+      throw Throwables.propagate(e);
+    }
+    this.metricStore = metricStore;
+    this.fetcherLimit = fetcherLimit;
+    this.metricsContextMap = Collections.<String, String>emptyMap();
+    processMetricsThreadList = new ArrayList<>();
+  }
+
+  @Override
+  public void setMetricsContext(MetricsContext metricsContext) {
+    this.metricsContext = metricsContext;
+    this.metricsContextMap = metricsContext.getTags();
+    metricStore.setMetricsContext(metricsContext);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running MessagingMetricsProcessorService");
+    MetricsConsumerMetaTable metaTable = getMetaTable();
+    if (metaTable == null) {
+      LOG.info("Could not get MetricsConsumerMetaTable, seems like we are being shut down");
+      return;
+    }
+
+    int i = 0;
+    while (i < metricsTopics.length && isRunning()) {
+      byte[] messageId = null;
+      try {
+        messageId = metaTable.getBytes(metricsTopicMetaKeys[i]);
+        LOG.info("Last processed MessageId for topic: {} is {}", metricsTopics[i], messageId);
```

Comment:
If there is no message id for the given topic, will the `getBytes` return `null` or throw exception?